src/main/kotlin/uk/gov/dluhc/printapi/service/CertificateBatchingService.kt

```diff
@@ -42,10 +42,7 @@ class CertificateBatchingService(
 
         return limitCertificates(certificatesPendingAssignment).chunked(batchSize).associate { batchOfCertificates ->
             val batchId = idFactory.batchId()
```

Comment:
Hi @Neil-Massey,

As we agreed, we can ignore the issue of exceeding the daily limit because of multiple print requests for the same certificate within the same batch. 

But, there is another issue. We can exceed the batch size limit for those certificates containing multiple print requests included in a full batch with 50 (batch-size) certificates. So for that batch, there will be more than 50 print request and the batch-size limit will be exceeded. In that case the batch might be rejected all together from the print provider.
 
I think this is a far more likely issue considering the batch size being 50.

I have a suggested solution like this:
<pre>
fun batchCertificates(): Map<String, List<Certificate>> {
        val certificatesPendingAssignment = certificateRepository.findByStatusOrderByApplicationReceivedDateTimeAsc(
            PENDING_ASSIGNMENT_TO_BATCH,
            Pageable.ofSize(maxUnBatchedRecords)
        )

        val printRequestsPendingAssignment = certificatesPendingAssignment.flatMap { it.printRequests }
            .filter { it.getCurrentStatus().status == PENDING_ASSIGNMENT_TO_BATCH }

        return limitPrintRequests(printRequestsPendingAssignment).chunked(batchSize).associate { batchOfPrintRequests ->
            val batchId = idFactory.batchId()
            batchId to batchOfPrintRequests.map { printRequest ->
                val certificate = certificatesPendingAssignment.first { it.printRequests.contains(printRequest) }
                certificate.addPrintRequestToBatch(printRequest, batchId)
                certificate
            }
        }
    }

    private fun limitPrintRequests(printRequestsPendingAssignmentToBatch: List<PrintRequest>): List<PrintRequest> {
        val startOfDay = Instant.now(clock).truncatedTo(DAYS)
        val endOfDay = startOfDay.plus(1, DAYS).minusSeconds(1)
        val countOfRequestsSentToPrintProvider =
            certificateRepository.getPrintRequestStatusCount(startOfDay, endOfDay, ASSIGNED_TO_BATCH)

        if ((printRequestsPendingAssignmentToBatch.size + countOfRequestsSentToPrintProvider) > dailyLimit) {
            logDailyLimit(printRequestsPendingAssignmentToBatch, countOfRequestsSentToPrintProvider, endOfDay)
            return printRequestsPendingAssignmentToBatch.subList(0, dailyLimit - countOfRequestsSentToPrintProvider)
        }

        return printRequestsPendingAssignmentToBatch
    }

    private fun logDailyLimit(
        printRequestsPendingAssignmentToBatch: List<PrintRequest>,
        countOfRequestsSentToPrintProvider: Int,
        endOfDay: Instant
    ) {
        val nextDay = endOfDay.plusSeconds(1)
        logger.warn {
            """Identified ${printRequestsPendingAssignmentToBatch.size} print requests to assign to a batch. 
            Daily print limit is $dailyLimit. 
            $countOfRequestsSentToPrintProvider print requests already sent to print provider today.
            Remaining capacity is ${dailyLimit - countOfRequestsSentToPrintProvider}.
            ${printRequestsPendingAssignmentToBatch.size + countOfRequestsSentToPrintProvider - dailyLimit} print requests won't be assigned to a batch until $nextDay at the earliest."""
        }
    }

</pre>

What do you think?

Reply:
Thanks for addressing it, @Neil-Massey.

Reply:
Agree - the limit should be applied to the PrintRequests and not the Certificates.  Have included the changes in this PR.